pkg/util/kubernetes/hpa/hpa.go

```diff
@@ -0,0 +1,389 @@
+func (c *HPAWatcherClient) Start() {
+	log.Info("Starting HPA Process ...")
+	tickerHPAWatchProcess := c.pollItl
+	tickerHPARefreshProcess := c.refreshItl
+
+	var resversion string
+	err := c.createConfigMapHPA()
+	if err != nil {
+		log.Errorf("Could not retrieve the ConfigMap %s to run the HPA process, stopping: %s", datadogHPAConfigMap, err.Error())
+		return
+	}
+
+	// Creating a leader election engine to make sure only the leader writes the metrics in the configmap and queries Datadog.
+	leaderEngine, err := leaderelection.GetLeaderEngine()
+	if err != nil {
+		log.Errorf("Could not ensure the leader election is running properly: %s", err)
+		return
+	}
+	leaderEngine.EnsureLeaderElectionRuns()
+
+	go func() {
+		for {
+			select {
+			// Ticker for the HPA Object watcher
+			case <-tickerHPAWatchProcess.C:
+				if !leaderEngine.IsLeader() {
+					continue
+				}
+				newHPA, modified, deleted, res, err := c.run(resversion)
+				if err != nil {
+					log.Errorf("Error while watching HPA Objects' activities: %s", err)
+					return
+				}
+				if res != resversion && res != "" {
+					resversion = res
+					if len(newHPA) > 0 {
+						c.storeHPA(newHPA)
+					}
+					if len(modified) > 0 {
+						c.storeHPA(modified)
+					}
+					if len(deleted) > 0 {
+						log.Infof("deleting if resver is diff")
+						c.removeEntryFromConfigMap(deleted)
+					}
+				}
+			// Ticker to run the refresh process for the stored external metrics
+			case <-tickerHPARefreshProcess.C:
+				if !leaderEngine.IsLeader() {
+					continue
+				}
+				// Updating the metrics against Datadog should not affect the HPA pipeline.
+				// If metrics are temporarily unavailable for too long, they will become `Valid=false` and won't be evaluated.
+				c.updateCustomMetrics()
+			}
+		}
+	}()
+}
+
+func (c *HPAWatcherClient) updateCustomMetrics() {
+	maxAge := int64(c.externalMaxAge.Seconds())
+
+	configMap, err := c.clientSet.CoreV1().ConfigMaps(c.ns).Get(datadogHPAConfigMap, metav1.GetOptions{})
+	if err != nil {
+		log.Infof("Error while retrieving the Config Map %s", datadogHPAConfigMap)
+		return
+	}
+
+	data := configMap.Data
+	if len(data) == 0 {
+		log.Debugf("No External Metrics to evaluate at the moment")
+		return
+	}
+
+	for _, d := range data {
+		cm := CustomExternalMetric{}
+		err := json.Unmarshal([]byte(d), &cm)
+		if err != nil {
+			log.Errorf("Could not unmarshal %#v", d)
+			continue
+		}
+		if metav1.Now().Unix()-cm.Timestamp > maxAge || !cm.Valid {
+			cm.Valid = false
+			cm.Timestamp = metav1.Now().Unix()
+
+			cm.Value, cm.Valid, err = c.validate(cm)
+			if err != nil {
+				log.Debugf("Could not update the metric %s from Datadog: %s", cm.Name, err.Error())
+				continue
+			}
+
+			c, err := json.Marshal(cm)
+			if err != nil {
+				log.Errorf("Error while marshalling the custom metric %s: %s", cm.Name, err.Error())
+				continue
+			}
+
+			data[cm.Name] = string(c)
+			cm.Valid = true // The Custom metric is ready to be consumed in the HPA pipeline
```

Comment:
I don't think setting `cm.Valid = true` is doing anything since you already marshalled it?

Reply:
True, I'm removing this one too https://github.com/DataDog/datadog-agent/pull/1883/files#diff-0fc3be6231b459e261c0d57460f1e200R306